evm/trie/database.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package trie
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/ChainSafe/log15"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+type Database struct {
+	path string
+	db   *leveldb.DB
+	lock sync.RWMutex
+	log  log15.Logger
+}
+
+type ProofDatabase struct {
+	db   map[string][]byte
+	lock sync.RWMutex
+}
+
+// NewProofDatabase returns a wrapped map
+func NewProofDatabase() *ProofDatabase {
+	db := &ProofDatabase{
+		db: make(map[string][]byte),
+	}
+
+	return db
+}
+
+func (db *ProofDatabase) Close() error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	db.db = nil
+	return nil
+}
+
+func (db *ProofDatabase) Has(key []byte) (bool, error) {
+	if db.db == nil {
+		return false, errors.New("database does not exist")
+	}
```

Comment:
These checks are unnecessary. As long as the constructor sets this up for you, it's up to the person using this code not to do stupid things